GLOSSARY.md

```diff
@@ -0,0 +1,167 @@
+
+## Binary connective
+There are sixteen logical connectives on two boolean variables. The most
+interesting in this context are listed below. There is also a truth table given
+which encodes the possible results of the operations like this
+
+```
+f(false, false) f(false, true) f(true, false) f(true, true).
+```
+
+As a shorthand the resulting truth table is encoded with `true = 1` and `false =
+0`.
+
+| Truth table | Operator symbol | Common name                      |
+|-------------|-----------------|----------------------------------|
+| 0001        | &&              | Conjunction; and                 |
+| 1001        | <=>             | Equivalence; if and only if; iff |
+| 1101        | =>              | Implication; if ... then         |
+
+## Binder
+A binder is an expression that binds a literal to a certain expression.
+Examples for binders:
+
+- The universal quantifier `forall(a)` states that a certain condition holds for
+  all allowed values for `a`.
+- A function definition `f(x) = a * x` is a binder for the variable `x` whereas
+  `a` is a free variable.
+- A sum `\sum_n x_n` binds the index variable `n`.
+
+## Canonical Form
+A formula in canonical form has the property that its DeBruijn indices are
+minimized. For example when the formula `forall<0, 1> { 0: A && 1: B }` is
+processed, both "branches" `0: A` and `1: B` are processed individually. The
+first branch would be in canonical form, the second branch not since the
+occurring DeBruijn index `1` could be replaced with `0`.
+
+## Clause
+In the A clause is the disjunction of several expressions. For example the clause
+`condition_1 || condition_2 || ...` states that at least one of the conditions
+holds.
+
+There are two notable special cases of clauses. A *Horn clause* has at most one
+positive literal. A *Definite clause* has exactly one positive literal.
+
+*Horn clauses* can be written in the form `A || !B || !C || ...` with `A` being
+the optional positive literal. Due to the equivalence `(P => Q) <=> (!P || Q)`
+the clause can be expressed as `B && C && ... => A` which means that A is true
+if `B`, `C`, etc. are all true. All rules in chalk are in this form. For example
+
+```
+struct A<T> {}
+impl<T> B for A<T> where T: C + D {}
+```
+
+is expressed as the *Horn clause* `(T: C) && (T: D) => (A<T>: B)`. This formula
+has to hold for all values of `T`. The second example
+
+```
+struct A {}
+impl B for A {}
+impl C for A {}
+```
+
+is expressed as the *Horn clause* `(A: B) && (A: C)`. Note the missing
+consequence.
+
+## Datum
+(empty)
+
+## DeBruijn Index
+DeBruijn indices are used to numerate free and bound literals inside a formula
+in an unambiguous way in order to work with numbers instead of the names of the
+literals. Given the example `forall<T> { exists<U> { T: Foo<Item=U> } }` the
+literal names `U` and `T` are replaced with `0` and `1` respectively: `forall<0>
+{ exists<1> { 0: Foo<Item=1> } }`.
```

Comment:
Maybe: "More generally, the debruijn index is the index of the binder where a name was defined, starting from the innermost binder and working out."